text/0000-infered-types.md

```diff
@@ -0,0 +1,174 @@
+
+# Motivation
+[motivation]: #motivation
+
+Using large libraries usually requires developers to import large amounts of traits, structures, and enums. To just call a function on an implementation can take upwards of 3 imports. One way developers have solved this is by importing everything from specific modules. Importing everything has its own problems like trying to guess where imports are actually from. Developers need a low compromise solution to solve this problem and that comes in the form of inferred types.
+
+Swift has had a system to infer types since around 2014. Its system has been praised by many developers around the world. It’s system is as follows:
+```swift
+enum EnumExample {
+   case variant
+   case variant2
+}
+
+
+example(data: .variant);
+```
+
+This RFC’s intent was to create something similar to the system already tried and tested in swift. Additionally, the underscore is already used to imply type’s lifetimes so it runs consistent with the rust theme.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+
+Implied types basically replace the path to a type with `_`. They can be used in places where strong typing exists. This includes function calls and function returns. You can think about the `_` as expanding into the type’s name.
+
+Function calls (struct):
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+// my_function(MyStruct {
+//     value: 1
+// });
+my_function(_ {
+   value: 1
+});
+```
+
+Function calls (impl):
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+// my_function(MyStruct::new()});
+my_function(_::new());
+```
+
+Function returns (enum):
+```rust
+fn my_function() -> MyEnum {
+   // MyEnum::MyVarient
+   _::MyVarient
+}
+```
+
+Match arms:
+```rust
+match Example::One {
+   _::One => println!("One!"),
+   _::Two => println!("Two!")
+};
+```
+
+It is important to note that `_` only represents the type; if you have (for example) another enum that can be coerced into the type, you will need to specify it manually. Additionally, any traits required to call an impl will still have to be imported.
+
+```rust
+fn my_function(data: MyStruct) { /* ... */ }
+
+
+my_function(MyStruct2::do_something().into()); // ✅
+
+
+my_function(_::do_something().into()); // ❌ variant or associated item not found in `MyStruct`
+```
+
+When developing, an IDE will display the type’s name next to the underscore as an IDE hint similar to implied types. Reading and maintaining code should not be impacted because the function name should give context to the type.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+This RFC should not take much to implement as you can think of the `_` as if it was just expanding to a type based on the context and where it’s used. Anywhere with a strongly typed argument as mentioned above can be inferred. This additionally includes let and const statements with type definitions in the left hand side.
```

Comment:
> Anywhere with a strongly typed argument as mentioned above can be inferred.

Mentioning this again, this is a bit vague.

Reply:
I think there are examples below.

Reply:
1. I think a *reference-level explanation* should be reference-like, not just have some examples.
2. These examples aren't even exhaustive :sweat_smile: